src/server/pfs/server/driver.go

```diff
@@ -391,102 +394,145 @@ func (d *driver) deleteAllRepos(ctx context.Context) ([]*pfs.Repo, error) {
 
 func (d *driver) deleteRepo(txnCtx *txncontext.TransactionContext, repo *pfs.Repo, force bool) error {
 	repos := d.repos.ReadWrite(txnCtx.SqlTx)
-
-	// check if 'repo' is already gone. If so, return that error. Otherwise,
-	// proceed with auth check (avoids awkward "access denied" error when calling
-	// "deleteRepo" on a repo that's already gone)
-	var repoInfo pfs.RepoInfo
-	err := repos.Get(repo, &repoInfo)
-	if err != nil {
+	repoInfo := &pfs.RepoInfo{}
+	if err := repos.Get(repo, repoInfo); err != nil {
 		if !col.IsErrNotFound(err) {
 			return errors.Wrapf(err, "error checking whether %q exists", repo)
 		}
+		return nil
 	}
-
-	// Check if the caller is authorized to delete this repo
-	if err := d.env.AuthServer.CheckRepoIsAuthorizedInTransaction(txnCtx, repo, auth.Permission_REPO_DELETE); err != nil {
-		return errors.EnsureStack(err)
+	if ok, err := d.canDeleteRepo(txnCtx, repo); err != nil {
+		return err
+	} else if !ok {
+		return nil
 	}
-
-	if !force {
-		if _, err := d.env.GetPPSServer().InspectPipelineInTransaction(txnCtx, pps.RepoPipeline(repo)); err == nil {
-			return errors.Errorf("cannot delete a repo associated with a pipeline - delete the pipeline instead")
-		} else if err != nil && !errutil.IsNotFoundError(err) {
-			return errors.EnsureStack(err)
+	related, err := d.relatedRepos(txnCtx, repo)
+	if err != nil {
+		return err
+	}
+	var bis []*pfs.BranchInfo
+	for _, ri := range related {
+		bs, err := d.listRepoBranches(txnCtx, ri)
+		if err != nil {
+			return err
 		}
+		bis = append(bis, bs...)
 	}
-	// if this is a user repo, delete any dependent repos
-	if repo.Type == pfs.UserRepoType {
-		var dependentRepos []pfs.RepoInfo
-		var otherRepo pfs.RepoInfo
-		if err := repos.GetByIndex(pfsdb.ReposNameIndex, pfsdb.ReposNameKey(repo), &otherRepo, col.DefaultOptions(), func(key string) error {
-			if otherRepo.Repo.Type != repo.Type {
-				dependentRepos = append(dependentRepos, otherRepo)
-			}
-			return nil
-		}); err != nil && !col.IsErrNotFound(err) {
-			return errors.Wrapf(err, "error finding dependent repos for %q", repo.Name)
+	if err := d.deleteBranches(txnCtx, bis, force); err != nil {
+		return err
+	}
+	for _, ri := range related {
+		if err := d.deleteRepoInfo(txnCtx, ri); err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		// we expect potentially complicated provenance relationships between dependent repos
-		// deleting all branches at once allows for topological sorting, avoiding deletion order issues
-		if err := d.deleteAllBranchesFromRepos(txnCtx, append(dependentRepos, repoInfo), force); err != nil {
-			return errors.Wrap(err, "error deleting branches")
+func (d *driver) deleteBranches(txnCtx *txncontext.TransactionContext, branchInfos []*pfs.BranchInfo, force bool) error {
+	sort.Slice(branchInfos, func(i, j int) bool { return len(branchInfos[i].Provenance) < len(branchInfos[j].Provenance) })
```

Comment:
It would probably be clearer to just reverse the sort and iterate forward through the branch infos.